lib/rules/no-deprecated.js

```diff
@@ -79,17 +79,17 @@ module.exports = {
       // 16.3.0
       deprecated.componentWillMount = [
         '16.3.0',
-        'constructor',
+        'UNSAFE_componentWillMount',
         'https://reactjs.org/docs/react-component.html#unsafe_componentwillmount'
       ];
       deprecated.componentWillReceiveProps = [
         '16.3.0',
-        'getDerivedStateFromProps',
+        'UNSAFE_componentWillReceiveProps',
```

Comment:
Why not suggest getDerivedStateFromProps? And why not lint against UNSAFE_ lifecycle methods?

Reply:
Because that's only one of the use cases for `componentWillReceiveProps`, so it's not always the right recommendation.

Reply:
What's the point of linting against `componentWillReceiveProps` if we're just going to recommend `UNSAFE_componentWillReceiveProps` anyway? Both are equally bad, no? They do the exact same thing, just one has a scary name.

Reply:
So that a human can look at it and make a human judgement about which to use.

Reply:
@sahrens, legacy lifecycles (including `componentWillReceiveProps`) will continue to work until version 17, however you will see deprecation warnings for them. In version 17, it will still be possible to use them, but they will be aliased with an “UNSAFE_”, see [details](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#component-lifecycle-changes).

Reply:
To clarify, you don't think we should lint against using any UNSAFE_ methods, even if we don't give a specific recommendation of what to use instead? Seems like linting against them earlier rather than later would help folks track down and remove usage before they start firing a bunch of deprecation warnings.

Reply:
Perhaps it would make more sense to make a new "strict-mode" lint rule? Any plans in that direction?

Reply:
I think that `no-deprecated` can and will warn against them, but not until *every* use case of them has a replacement (this is not currently the case).

Reply:
There're no deprecation warnings regarding using `UNSAFE_` methods in strict mode, there is unsafe lifecycle methods warning: 
![unsafe lifecycle methods warning](https://reactjs.org/static/strict-mode-unsafe-lifecycles-warning-e4fdbff774b356881123e69ad88eda88-2535d.png)

I'm not sure regarding `strict-mode` rule, I vote for adding `no-unsafe` rule that will be more specific. In the same way that it was done for `no-string-refs` rule -- using string ref API also cause warnings in strict mode. @ljharb, I could take a look at it if you find it reasonable.

Reply:
Modes are not a conceptual pattern I'd want to encourage, "strict" or otherwise.

I don't think a "no-unsafe" rule makes sense when for users of React < 16, it is perfectly safe, and for users of React >= 17, the methods won't work at all.

The purpose of these deprecations from React is to encourage people to migrate to replacements as early as possible, so the breaking change in v17 that removes the old ones will be as painless as possible. When React 17 comes out, we'll certainly lint against invalid lifecycle methods, including newly-invalid ones - but I'm not sure how much value we'd get out of warnings in React 16 outside of `no-deprecated`.

Reply:
>I don't think a "no-unsafe" rule makes sense when for users of React < 16, it is perfectly safe, and for users of React >= 17, the methods won't work at all.

It's not true, even in version 17, it will still be possible to use UNSAFE_ methods, see [Component Lifecycle Changes](https://reactjs.org/blog/2018/03/29/react-v-16-3.html#component-lifecycle-changes).

Reply:
In that case, a separate "no-unsafe" rule, that only focuses on UNSAFE_ methods, seems reasonable.